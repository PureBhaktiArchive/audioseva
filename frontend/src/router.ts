/*
 * sri sri guru gauranga jayatah
 */

import firebase from "firebase/app";

import Vue from "vue";
import Router from "vue-router";

Vue.use(Router);

export const router = new Router({
  mode: "history",
  routes: [
    { path: "*", redirect: "/" },
    {
      path: "/login",
      meta: { guestOnly: true },
      component: () => import("@/views/Login.vue")
    },
    {
      path: "/",
      meta: { requireAuth: true },
      component: () => import("@/views/Dashboard.vue"),
      children: [
        {
          path: "",
          component: () => import("@/views/Home.vue")
        },
        {
          path: "cr/allot",
          component: () => import("@/views/CRAllotment.vue")
        },
        {
          path: "sqr/",
          component: () => import("@/views/SQR.vue"),
          meta: { activator: true, activatorName: "Sound Quality Reporting", menuIcon: "fas fa-headphones" },
          children: [
            { path: "",
              component: () => import("@/views/SQRFiles.vue"),
              meta: { menuItem: true}
            },
            {
              path: "allot",
              component: () => import("@/views/SQRAllotment.vue"),
              meta: { menuItem: true }
            },
            { path: "statistics",
              component: () => import("@/views/SQRFileStatistics.vue"),
              meta: { menuItem: true }
            }
          ]
        },
        {
          path: "se/",
          component: () => import("@/views/SE/SE.vue"),
          meta: { activator: true, activatorName: "Sound Engineering", menuIcon: "fas fa-music" },
          children: [
            { path: "", component: () => import("@/views/SE/Tasks.vue"), meta: { menuItem: true } },
            { path: "allot", component: () => import("@/views/SE/Allotment.vue"), meta: { menuItem: true } }
          ]
        },
        {
          path: "te/allot",
          component: () => import("@/views/TEAllotment.vue")
        },
        {
          path: "te/fc/allot",
          component: () => import("@/views/TFCAllotment.vue")
        },
        {
          path: "/qc/allot",
          component: () => import("@/views/QCAllotment.vue")
        }
      ]
    }
  ]
});

router.beforeEach((to, from, next) => {
  let currentUser = firebase.auth().currentUser;
  let requireAuth = to.matched.some(record => record.meta.requireAuth);
  let guestOnly = to.matched.some(record => record.meta.guestOnly);

  if (requireAuth && !currentUser)
    next({
      path: "/login",
      query: { redirect: to.fullPath }
    });
  else if (guestOnly && currentUser) next("/");
  else next();
});
