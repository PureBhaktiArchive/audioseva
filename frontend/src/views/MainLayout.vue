<template>
  <base-layout>
    <template slot="aboveContent">
      <v-navigation-drawer fixed :clipped="$vuetify.breakpoint.lgAndUp" v-model="sidebar" app v-if="currentUser">
        <v-toolbar flat class="transparent">
          <v-list class="pa-0">
            <v-list-tile avatar>
              <v-list-tile-avatar>
                <img :src="currentUser.photoURL">
              </v-list-tile-avatar>
              <v-list-tile-content>
                <v-list-tile-title>{{currentUser.displayName}}</v-list-tile-title>
              </v-list-tile-content>
              <v-list-tile-action>
                <v-btn icon ripple @click="signOut">
                  <v-icon>fas fa-sign-out-alt</v-icon>
                </v-btn>
              </v-list-tile-action>
            </v-list-tile>
          </v-list>
        </v-toolbar>
        <v-list dense expand>
          <v-divider></v-divider>
          <template v-for="(item, index) in routes">
            <v-list-group
              :key="index"
              v-if="item.meta && item.meta.activator"
              no-action
              :prepend-icon="item.meta.menuIcon"
              :value="index === 0"
            >
              <v-list-tile slot="activator">
                <v-list-tile-content>
                  <v-list-tile-title>{{ item.meta.activatorName }}</v-list-tile-title>
                </v-list-tile-content>
              </v-list-tile>
              <menu-links :parentRoute="item" :routes="item.children"></menu-links>
            </v-list-group>
            <v-list-tile :to="item.path" v-else-if="item.meta.menuItem" :key="`no-nested-${index}`">
              <v-list-tile-action>
                <v-icon>{{ item.meta.menuIcon }}</v-icon>
              </v-list-tile-action>
              <v-list-tile-content>
                <v-list-tile-title>
                  {{ item.meta.menuName }}
                </v-list-tile-title>
              </v-list-tile-content>
            </v-list-tile>
          </template>
          <v-divider></v-divider>
          <v-list-tile @click="signOut" v-if="currentUser">
            <v-list-tile-action>
              <v-icon left>fas fa-sign-out-alt</v-icon>
            </v-list-tile-action>
            <v-list-tile-content>SignOut</v-list-tile-content>
          </v-list-tile>
        </v-list>
      </v-navigation-drawer>

      <v-toolbar :clipped-left="$vuetify.breakpoint.lgAndUp" fixed app>
        <v-toolbar-side-icon @click="sidebar = !sidebar" v-if="currentUser">
        </v-toolbar-side-icon>
        <v-toolbar-title>
          <router-link to="/" tag="span" style="cursor: pointer">
            {{ appTitle }}
          </router-link>
        </v-toolbar-title>
      </v-toolbar>
    </template>
    <template slot="aboveRoute">
      <slot name="aboveRoute"></slot>
    </template>
  </base-layout>
</template>

<script lang="ts">
import { Component, Vue } from "vue-property-decorator";
import { mapState, mapActions } from "vuex";
import BaseLayout from "@/views/BaseLayout.vue";
import MenuLinks from "@/components/MenuLinks.ts";

@Component({
  name: "MainLayout",
  components: {
    BaseLayout,
    MenuLinks
  },
  computed: {
    ...mapState("user", ["currentUser"])
  },
  methods: {
    ...mapActions("user", ["signOut"])
  }
})
export default class MainLayout extends Vue {
  appTitle = "Audio Seva";
  navLinks = [];
  sidebar = false;

  mounted() {
    // @ts-ignore
    this.navLinks = this.$router.options.routes.find(
      (route: any) => route.path === "/"
    ).children;
  }

  get routes(): any {
    // @ts-ignore
    return this.navLinks.filter((route: any) => {
      return route.meta && (route.meta.activator || route.meta.menuItem);
    });
  }
}
</script>

<style scoped>
</style>
