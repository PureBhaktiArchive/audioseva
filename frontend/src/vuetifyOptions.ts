export default {
  icons: {
    iconfont: 'fa',
    values: {
      plus: 'fas fa-plus',
      listening: 'fas fa-headphones',
      track: 'fas fa-cut',
      sound: 'fas fa-music',
      download: 'fas fa-file-download',
      delete: 'fas fa-trash',
      event: 'fas fa-calendar',
      undo: 'fas fa-undo',
      upload: 'fas fa-upload',
      check: 'fas fa-check',
    },
  },
} as any;
