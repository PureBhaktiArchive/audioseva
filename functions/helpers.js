const functions = require('firebase-functions');
let SibApiV3Sdk = require('sib-api-v3-sdk');
const sendInBlueSecretKey = functions.config().audioseva.send_in_blue.key;
const defaultClient = SibApiV3Sdk.ApiClient.instance;
let apiKey = defaultClient.authentications['api-key'];
apiKey.apiKey = sendInBlueSecretKey;



exports.checkValidMP3 = filePath => (filePath.startsWith("mp3/") && filePath.endsWith(".mp3"))


exports.removeFromDB = (db, dbPath) => {
    let ref = db.ref(dbPath);
    ref.remove()
        .then(() => console.log("Deleted."))
        .catch(error => console.log(error));
}


exports.sendEmail = (to, bcc, templateId, params) => {
    let apiInstance = new SibApiV3Sdk.SMTPApi();
    sendEmail = new SibApiV3Sdk.SendSmtpEmail();

    sendEmail = {
        to: [{ email: to }], // add a name prop if it fails to send
        bcc,
        templateId,
        params
    };

    apiInstance.sendTransacEmail(sendEmail).then(function(data) {
        console.log('API called successfully. Returned data: ');
        console.log(data);
        return 1;
    }, err => console.log(err)).catch(err => console.log(err));
}
