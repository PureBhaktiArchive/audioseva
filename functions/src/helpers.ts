const functions = require('firebase-functions');
let SibApiV3Sdk = require('sib-api-v3-sdk');
const sendInBlueSecretKey = functions.config().send_in_blue.key;
const defaultClient = SibApiV3Sdk.ApiClient.instance;
let apiKey = defaultClient.authentications['api-key'];
apiKey.apiKey = sendInBlueSecretKey;



export const checkValidMP3 = filePath => (filePath.startsWith("mp3/") && filePath.endsWith(".mp3"))

export const createMP3DBRef = (filePath, _module) => {
    const parts = filePath.split('/');
        
    const list = parts[1];
    const mp3 = parts[2];
    let file_name = mp3.slice(0, -4);

    return `/${_module}/files/${list}/${file_name}`;
}

export const extractListFromFilename = (fileName) => {
    return fileName.match(/^[^-]*[^ -]/g)[0];
}


export const storeFileNameToDB = (filePath, db, _module) => {
    if(checkValidMP3(filePath)) {
        let ref = db.ref(createMP3DBRef(filePath, _module));

        //checking if the file already exists in the RT db
        ref.child("status").once('value')
        .then(snapshot => {
            if(!snapshot.exists())
                ref.set({status: "Spare"});
            return 1;
        }).catch(err => console.log(err));
    }
}

export const removeFromDB = (db, dbPath) => {
    let ref = db.ref(dbPath);
    ref.remove()
        .then(() => console.log("Deleted."))
        .catch(error => console.log(error));
}


export const sendEmail = (to, bcc, templateId, params) => {
    let apiInstance = new SibApiV3Sdk.SMTPApi();
    let sendEmail = new SibApiV3Sdk.SendSmtpEmail();

    sendEmail = {
        to: [{ email: to }], // add a name prop if it fails to send
        bcc,
        templateId,
        params
    };

    apiInstance.sendTransacEmail(sendEmail).then(function(data) {
        // Message sending result
        console.log(data);
        return 1;
    }, err => console.log(err)).catch(err => console.log(err));
}
