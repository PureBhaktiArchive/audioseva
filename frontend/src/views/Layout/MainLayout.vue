<template>
  <v-app>
    <v-navigation-drawer
      fixed
      :clipped="$vuetify.breakpoint.lgAndUp"
      v-model="sidebar"
      app
      v-if="currentUser"
      width="300px"
    >
      <v-list class="pa-0">
        <v-list-item>
          <v-list-item-avatar>
            <img :src="currentUser.photoURL" />
          </v-list-item-avatar>
          <v-list-item-content>
            <v-list-item-title>{{ currentUser.displayName }}</v-list-item-title>
          </v-list-item-content>
          <v-list-item-action>
            <v-btn color="rgba(0, 0, 0, 0.87)" icon ripple @click="signOut">
              <v-icon>fas fa-sign-out-alt</v-icon>
            </v-btn>
          </v-list-item-action>
        </v-list-item>
      </v-list>
      <v-list dense expand>
        <v-divider></v-divider>
        <template v-for="(item, index) in menuItems">
          <menu-links
            :parentRoute="item"
            :key="index"
            v-if="item.meta && item.meta.activator"
            :activeClass="getActiveClass(item.path)"
          >
          </menu-links>
          <v-list-item
            :to="`/${item.path}`"
            v-else-if="item.meta.menuItem"
            :key="`no-nested-${index}`"
            :active-class="getActiveClass(item.path)"
          >
            <v-list-item-icon>
              <v-icon v-text="item.meta.menuIcon"></v-icon>
            </v-list-item-icon>
            <v-list-item-content>
              <v-list-item-title>{{ item.meta.menuName }}</v-list-item-title>
            </v-list-item-content>
          </v-list-item>
        </template>
        <v-divider></v-divider>
        <v-list-item @click="signOut" v-if="currentUser">
          <v-list-item-action>
            <v-icon left>fas fa-sign-out-alt</v-icon>
          </v-list-item-action>
          <v-list-item-content>SignOut</v-list-item-content>
        </v-list-item>
      </v-list>
    </v-navigation-drawer>

    <v-app-bar :clipped-left="$vuetify.breakpoint.lgAndUp" fixed app>
      <v-app-bar-nav-icon
        @click="sidebar = !sidebar"
        v-if="currentUser"
      ></v-app-bar-nav-icon>
      <v-toolbar-title>
        <router-link to="/" tag="span" style="cursor: pointer">{{
          appTitle
        }}</router-link>
      </v-toolbar-title>
    </v-app-bar>
    <v-content>
      <v-container fluid>
        <v-breadcrumbs
          v-if="breadcrumbs.length"
          :items="breadcrumbs"
          divider=">"
        ></v-breadcrumbs>
        <router-view></router-view>
      </v-container>
    </v-content>
  </v-app>
</template>

<script lang="ts">
import { Component, Vue, Watch } from "vue-property-decorator";
import _ from "lodash";
import { mapState, mapActions } from "vuex";
import { RouteConfig } from "vue-router";
import { getMenuItems } from "@/router";
import MenuLinks from "@/components/MenuLinks";

const pathMap = {
  cr: "CR",
  sqr: "SQR",
  te: "TE",
  allot: "Allotment"
};

@Component({
  name: "MainLayout",
  components: {
    MenuLinks
  },
  computed: {
    ...mapState("user", ["currentUser", "roles"])
  },
  methods: {
    ...mapActions("user", ["signOut"])
  }
})
export default class MainLayout extends Vue {
  appTitle = "Audio Seva";
  sidebar = false;
  menuItems: RouteConfig[] = [];
  breadcrumbs: any[] = [];

  @Watch("roles", { immediate: true })
  handleUserRolesChange(newRoles: { [key: string]: any } | null) {
    if (newRoles) {
      this.menuItems = getMenuItems();
    } else {
      this.menuItems = [];
    }
  }

  @Watch("$route", { immediate: true, deep: true })
  getBreadcrumbs({ path, params, matched }: any) {
    const breadcrumbs: any[] = [];
    const auth = _.get(
      [...matched].reverse().find(({ meta }) => meta.auth),
      "meta.auth"
    );
    if (path !== "/" && auth && (auth.requireClaims || auth.requireAuth)) {
      const paths = path.split("/");
      const pathsLength = paths.length - 1;
      paths.forEach((item: string, index: number) => {
        breadcrumbs.push({
          text: this.getText(item, params),
          disabled: index === pathsLength,
          to: this.getPath(item, index, paths),
          exact: true
        });
      });
    }
    this.breadcrumbs = breadcrumbs;
  }

  getPath(item: string, index: number, paths: string[]) {
    return `${paths.slice(0, index).join("/")}/${item}`;
  }

  getText(path: string, params: any) {
    if (path) {
      let customText = _.get(
        pathMap,
        path,
        path[0].toUpperCase() + path.substring(1)
      );
      if (typeof customText === "function") {
        customText = customText(params);
      }
      return customText;
    }
    return "Home";
  }

  getActiveClass(path: string) {
    return this.$route.path.includes(path.substring(0, path.length - 1))
      ? "primary--text"
      : "inactive-menu";
  }
}
</script>

<style scoped>
>>> .inactive-menu {
  color: rgba(0, 0, 0, 0.87);
}

>>> .v-content__wrap {
  display: flex;
}
</style>
