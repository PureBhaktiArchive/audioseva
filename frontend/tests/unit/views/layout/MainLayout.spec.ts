import { router } from '@/router';
import MainLayout from '@/views/Layout/MainLayout.vue';
import vuetifyOptions from '@/vuetifyOptions';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import flushPromises from 'flush-promises';
import Vue from 'vue';
import VueRouter from 'vue-router';
import Vuetify from 'vuetify';
import { mockClaims } from '../../helpers';

describe('MainLayout', () => {
  let localVue: typeof Vue;

  beforeEach(() => {
    localVue = createLocalVue();
    localVue.use(VueRouter);
  });

  test.each`
    claims
    ${{ TE: { editor: true } }}
    ${{ SQR: { coordinator: true } }}
  `(
    'should render menu items that match claims $claims',
    async ({ claims }) => {
      await mockClaims(claims);
      const wrapper = shallowMount(MainLayout, {
        localVue,
        router,
        vuetify: new Vuetify(vuetifyOptions),
        computed: {
          currentUser() {
            return {};
          },
          roles() {
            return claims;
          },
        },
      });
      await flushPromises();
      expect((wrapper.vm as any).menuItems).toMatchSnapshot();
    }
  );
});
