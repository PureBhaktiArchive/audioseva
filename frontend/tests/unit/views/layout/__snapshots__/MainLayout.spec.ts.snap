// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainLayout should render menu items that match claims {"TE": true} 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "component": [Function],
        "meta": Object {
          "auth": Object {
            "requireClaims": Object {
              "TE": true,
            },
          },
          "menuItem": true,
          "menuLinkName": "My Tasks",
        },
        "path": "my",
      },
      Object {
        "component": [Function],
        "meta": Object {
          "auth": Object {
            "requireClaims": Object {
              "TE": true,
              "coordinator": true,
            },
          },
          "homePageLink": Object {
            "text": "Upload",
          },
          "menuItem": true,
        },
        "path": "upload",
      },
      Object {
        "component": [Function],
        "meta": Object {
          "auth": Object {
            "requireClaims": Object {
              "TE": true,
              "coordinator": true,
            },
          },
          "menuItem": false,
        },
        "path": "tasks/:taskId",
      },
    ],
    "component": [Function],
    "meta": Object {
      "activator": true,
      "activatorName": "Track Editing",
      "auth": Object {
        "requireClaims": Object {
          "coordinator": true,
        },
      },
      "menuIcon": "fas fa-cut",
    },
    "path": "te/",
  },
]
`;

exports[`MainLayout should render menu items that match claims {"coordinator": true} 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "component": [Function],
        "meta": Object {
          "menuItem": true,
        },
        "path": "allot",
      },
    ],
    "component": [Function],
    "meta": Object {
      "activator": true,
      "activatorName": "Sound Quality Reporting",
      "auth": Object {
        "requireClaims": Object {
          "coordinator": true,
        },
      },
      "menuIcon": "fas fa-headphones",
    },
    "path": "sqr/",
  },
  Object {
    "children": Array [
      Object {
        "component": [Function],
        "meta": Object {
          "homePageLink": Object {
            "text": "TE",
          },
          "menuItem": true,
        },
        "path": "tasks",
      },
      Object {
        "component": [Function],
        "meta": Object {
          "menuItem": true,
        },
        "path": "allot",
      },
      Object {
        "component": [Function],
        "meta": Object {
          "auth": Object {
            "requireClaims": Object {
              "TE": true,
              "coordinator": true,
            },
          },
          "homePageLink": Object {
            "text": "Upload",
          },
          "menuItem": true,
        },
        "path": "upload",
      },
      Object {
        "component": [Function],
        "meta": Object {
          "auth": Object {
            "requireClaims": Object {
              "TE": true,
              "coordinator": true,
            },
          },
          "menuItem": false,
        },
        "path": "tasks/:taskId",
      },
    ],
    "component": [Function],
    "meta": Object {
      "activator": true,
      "activatorName": "Track Editing",
      "auth": Object {
        "requireClaims": Object {
          "coordinator": true,
        },
      },
      "menuIcon": "fas fa-cut",
    },
    "path": "te/",
  },
]
`;
